### PR TITLE
[KYLIN-4939] Transform lookup table snapshot from segment level to cu…

### PR DESCRIPTION
**Pull request by Ted-Jiang**
…be level

## Proposed changes

We find many use cases should use cube-level lookup table snapshot instead of segment-level ones. For segment-level ones, it will cause many issues. For example for table BATCH_VIEWS.xxx, it changes everyday. Now there're around 2000 snapshots.

For building a new snapshot, it will check the duplication to avoid duplicated snapshots, which means it will check all of the 2000 snapshots. It really costs time.
For query & building, to cache the snapshot, if each takes around 6MB, the total will need around 12GB memory.
After this change, we can not only speedup cubing but also reduce the burden of memory cost.

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have create an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [ ] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this change need a document change, I will prepare another pr against the `document` branch
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin or dev@kylin by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: core-cube/src/main/java/org/apache/kylin/cube/CubeDescManager.java [hunk 1]
```diff
@@ -256,6 +256,18 @@ public CubeDesc updateCubeDesc(CubeDesc desc) throws IOException {
         }
     }
 
+    public CubeDesc updateCubeDescToBeGlobal(CubeDesc desc, String lookupTable, boolean global) throws IOException {
```
Comment (hit-lacus) [inline core-cube/src/main/java/org/apache/kylin/cube/CubeDescManager.java:259]: The name of this method is misleading.

File: core-cube/src/main/java/org/apache/kylin/cube/model/CubeDesc.java [hunk 2]
```diff
@@ -623,6 +624,12 @@ public boolean consistentWith(CubeDesc another) {
         return this.calculateSignature().equals(another.calculateSignature());
     }
 
+    public CubeDesc lastedCopyForWrite() {
```
Comment (hit-lacus) [inline core-cube/src/main/java/org/apache/kylin/cube/model/CubeDesc.java:627]: Typo, "lasted" or "latest" ?

File: server-base/src/main/java/org/apache/kylin/rest/controller/CubeController.java [hunk 3]
```diff
@@ -346,6 +346,35 @@ public JobInstance rebuildLookupSnapshot(@PathVariable String cubeName,
...
+        List<CubeInstance> result = new ArrayList<>();
+
+        final CubeManager cubeMgr = cubeService.getCubeManager();
+        String[] changeCubes = cubeNames.toUpperCase(Locale.ROOT).split(",");
+        for (String cubeName : changeCubes) {
+            try {
+                checkCubeExists(cubeName);
+                final CubeInstance cube = cubeMgr.getCube(cubeName);
+                CubeInstance cubeInstance = cubeService.changeLookupSnapshotBeGlobal(cube, tableName);
+                logger.info("cube %s change snapshotTable %s global Success", cubeName, tableName);
```
Comment (hit-lacus) [inline server-base/src/main/java/org/apache/kylin/rest/controller/CubeController.java:368]: Placeholder should be `%` or `{}` ?

Comment (Ted-Jiang) [reply server-base/src/main/java/org/apache/kylin/rest/controller/CubeController.java]: fix

File: core-cube/src/main/java/org/apache/kylin/cube/model/CubeDesc.java (hunk 2 above)
Comment (Ted-Jiang) [reply core-cube/src/main/java/org/apache/kylin/cube/model/CubeDesc.java]: thx!

**Review by hit-lacus — APPROVED** [review]
Approved

**Pull request merged**